### PR TITLE
Use KeyStore instead of raw keys with CAs

### PR DESCRIPTION
**Pull request by nklaassen**
This PR updates all uses of private keys held in CAs to use the `KeyStore` interface instead of interacting with the private keys directly, so that the keys may be held in an HSM.

Currently, only the `rawKeyStore` implementation is used (which uses raw in-memory keys with no HSM), so this should result in no functional change.

Remaining steps to implement for full HSM support:
1. Each Auth server with a configured HSM should add its own private keys to each CA.
2. Support for CA rotation with multiple auth servers.
3. User configuration of HSM.

### CODE REVIEW

File: lib/auth/auth.go [hunk 1]
```diff
@@ -290,6 +297,9 @@ type Server struct {
 	// streamer is events sessionstreamer, used to create continuous
 	// session related streams
 	streamer events.Streamer
+
+	// keyStore is an interface for interacting with private keys
```
Comment (tcsc) [inline lib/auth/auth.go:301]: I'd expand this comment a bit - whose private keys, what scope, etc.

**Review by tcsc — APPROVED** [review]

**Review by xacrimon — APPROVED** [review]

File: lib/client/keystore_test.go [hunk 2]
```diff
@@ -428,7 +428,7 @@ func newSelfSignedCA(privateKey []byte) (*tlsca.CertAuthority, auth.TrustedCerts
 	if err != nil {
 		return nil, auth.TrustedCerts{}, trace.Wrap(err)
 	}
-	ca, err := tlsca.FromKeys(cert, privateKey)
+	ca, err := tlsca.FromCertAndSigner(cert, rsaKey.(*rsa.PrivateKey))
```
Comment (r0mant) [inline lib/client/keystore_test.go:431]: Check type assertion or is this guaranteed not to panic?

File: lib/auth/rotate.go [hunk 3]
```diff
@@ -380,7 +379,7 @@ func (a *Server) autoRotate(ca types.CertAuthority) error {
 
 // processRotationRequest processes rotation request based on the target and
 // current phase and state.
-func processRotationRequest(req rotationReq) (types.CertAuthority, error) {
+func processRotationRequest(req rotationReq, keyStore keystore.KeyStore) (types.CertAuthority, error) {
```
Comment (r0mant) [inline lib/auth/rotate.go:382]: Nit: Would it make sense to just make this method and startNewRotation members of Server so you don't have to pass an additional function argument?

File: lib/auth/auth.go [hunk 4]
```diff
@@ -497,24 +501,8 @@ func (a *Server) GenerateHostCert(hostPublicKey []byte, hostID, nodeName string,
...
-		}
-		signer, err := ssh.ParsePrivateKey(kp.PrivateKey)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		signer = sshutils.AlgSigner(signer, sshutils.GetSigningAlgName(ca))
-		return signer, nil
-	}
-	return nil, trace.NotFound("no raw SSH private key found in CA for %q", ca.GetClusterName())
+func (a *Server) GetKeyStore() keystore.KeyStore {
```
Comment (r0mant) [inline lib/auth/auth.go:504]: Nit: Missing godoc.

**Review by r0mant — APPROVED** [review]

File: lib/auth/auth.go (hunk 1 above)
Comment (nklaassen) [reply lib/auth/auth.go]: I've elaborated a bit here

File: lib/auth/rotate.go (hunk 3 above)
Comment (nklaassen) [reply lib/auth/rotate.go]: good point, done in latest commit

**Pull request merged**